src/Fixer/PhpUnit/PhpUnitTestCaseStaticMethodCallsFixer.php

```diff
@@ -0,0 +1,449 @@
+        'stringEndsWith' => true,
+        'stringStartsWith' => true,
+
+        // TestCase methods
+        'any' => true,
+        'at' => true,
+        'atLeast' => true,
+        'atLeastOnce' => true,
+        'atMost' => true,
+        'exactly' => true,
+        'never' => true,
+        'onConsecutiveCalls' => true,
+        'once' => true,
+        'returnArgument' => true,
+        'returnCallback' => true,
+        'returnSelf' => true,
+        'returnValue' => true,
+        'returnValueMap' => true,
+        'setUpBeforeClass' => true,
+        'tearDownAfterClass' => true,
+        'throwException' => true,
+    ];
+
+    private $conversionMap = [
+        self::CALL_TYPE_THIS => [[T_OBJECT_OPERATOR, '->'], [T_VARIABLE, '$this']],
+        self::CALL_TYPE_SELF => [[T_DOUBLE_COLON, '::'], [T_STRING, 'self']],
+        self::CALL_TYPE_STATIC => [[T_DOUBLE_COLON, '::'], [T_STATIC, 'static']],
+    ];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Calls to `PHPUnit\Framework\TestCase` static methods must all be of the same type, either `$this->`, `self::` or `static::`.',
+            [
+                new CodeSample(
+                    '<?php
+final class MyTest extends \PHPUnit_Framework_TestCase
+{
+    public function testMe()
+    {
+        $this->assertSame(1, 2);
+        self::assertSame(1, 2);
+        static::assertSame(1, 2);
+    }
+}
+'
+                ),
+            ],
+            null,
+            'Risky when PHPUnit methods are overridden or not accessible, or when project has PHPUnit incompatibilities.'
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAllTokenKindsFound([T_CLASS, T_STRING]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isRisky()
+    {
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach (array_reverse($this->findPhpUnitClasses($tokens)) as $indexes) {
+            $this->fixPhpUnitClass($tokens, $indexes[0], $indexes[1]);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function createConfigurationDefinition()
+    {
+        $thisFixer = $this;
+
+        return new FixerConfigurationResolver([
+            (new FixerOptionBuilder('call_type', 'The call type to use for referring to PHPUnit methods.'))
+                ->setAllowedTypes(['string'])
+                ->setAllowedValues(array_keys($this->allowedValues))
+                ->setDefault('static')
+                ->getOption(),
+            (new FixerOptionBuilder('methods', 'Dictionary of `method` => `call_type` values that differ from the default strategy.'))
+                ->setAllowedTypes(['array'])
+                ->setAllowedValues([static function ($option) use ($thisFixer) {
```

Comment:
I agree that sounds verbose, but I find it more clear about the intentions.
Here I explicitly tell that this function will never be bound to anything but the fixer.
Otherwise someone in the future may change the binding (for example in `setAllowedValues`) and s**t happens.

Reply:
Well, tests are here to prevent s**t from happening aren't they :p

Reply:
Language constructs are stronger than tests, but I agree that a standard should be preferred over a perfect, but isolated, solution.

Reply:
someone can always rebind sth, and having that protection here is IMO not worth it, if we don't protect ourselves everywhere else. So either we forbid to rebind on language level, having all lambdas static, or we forbid to do so by "good code rule" of "don't rebind".
Here, aliasing `$this` to pass an alias instead of using value directly is confusing, imho.

Reply:
Ok, I'll revert it tomorrow

Reply:
Can't do that: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/src/FixerConfiguration/FixerOption.php#L77

Reply:
That's code introduced by you @julienfalque :smile:

Reply:
Damn, forgot about that one. Yep, using closures bound to a fixer for configuration caused big memory leaks. This is probably still true so you actually need to keep the closure static :/

Reply:
If can be any help I can import within the `use` only the variables I use, but i'm not an expert in PHP memory leaks / gc, so don't know if would be an improvement or a degradation.

Reply:
The memory leak is caused when the closure is bound to the fixer it is declared in:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/852339dc22539ea5d4ef5be0c1f83b4db3ec43b1/src/FixerConfiguration/FixerOption.php#L153-L179

so your current code is fine.